### ISSUE TITLE
TypeScript non-null-assertion-operator returning undefined

### ISSUE THREAD

**Issue by domgaulton** [DevExpress/devextreme-reactive#2818]
https://github.com/DevExpress/devextreme-reactive/blob/69eab29acbbec9c691890afffd24cddc55c3248f/packages/dx-react-chart/src/utils/series-helper.tsx#L71

I don't actually know TypeScript but keep getting an error on line 71 of series-helper.tsx so had a look around and saw `valScale: scales[getValueDomainName(currentSeries!.scaleName)],` Initially I thought the `!` in the middle of currentSeries and scaleName was weird but did some reading and found that it was a `non-null-assertion-operator`. From my understanding this is used when the object will have a key so in this case `currentSeries` **WILL** have a key value for `scaleName`. 

https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-0.html#non-null-assertion-operator

The problem is that my code keeps crashing at this point saying
```
TypeError: Cannot read property 'scaleName' of undefined
(anonymous function)
src/utils/series-helper.tsx:71
  68 | const currentSeries = findSeriesByName(symbolName, series);
  69 | const currentScales: Scales = {
  70 |   argScale: scales[ARGUMENT_DOMAIN],
> 71 |   valScale: scales[getValueDomainName(currentSeries!.scaleName)],
     | ^  72 | };
  73 | const Path: React.ComponentType<PathComponentProps> =
  74 |   currentSeries.seriesComponent as any;
```

This is causing me lots of problems so I wondered if anyone else had seen this and if it was an issue?

**Comment by AryamnovEugeniy** [comment]
Hello @domgaulton !

Could you, please, provide us with an example where the issue is reproduced? In this case we will be able to help you sooner. For example, you may update [this example](https://codesandbox.io/s/n12e8).

**Comment by domgaulton** [comment]
Hi @AryamnovEugeniy,

The application that I am building uses API calls etc so I have instead tried to replicate it in the sandbox.

What my application does is send the graph a single prop (in the case of the sandbox see `data1` `data2` or `data3`. The idea is that the prop object gets sent as one object so all data syncs up and works at the same time. I also build the bar items and stacks dynamically as these can change based on filters (not included).

Within the sandbox please click the change button in the top right - currently there are only 3 data points but I am getting an error which I thought perhaps was linked. It's not the same error but perhaps this can offer some insight?

https://codesandbox.io/s/funny-wood-hdh6i?file=/demo.js:347-353

Thanks

Dom Gaulton

**Comment by AryamnovEugeniy** [comment]
Thank you for the example you provided! I managed to reproduce the behaviour you described in your first comment in your sample. I will update this issue as soon as we have any results.

**Comment by AryamnovEugeniy** [comment]
We need time to investigate this issue further but I found a workaround: https://codesandbox.io/s/blissful-bird-jr4po.
I provide a new `key` to the `Chart` every time the number of `barItems` changes.

**Comment by AryamnovEugeniy** [comment]
Hello once again @domgaulton,

Could you tell me if this workaround was helpful to you?

**Comment by domgaulton** [comment]
Hi @AryamnovEugeniy 

Yes this has worked as a temporary fix - thank you. Great to be able to add animation, hover effects and tooltips back in and see it working. 

Thanks for your speedy response too!

Dom

**Comment by AryamnovEugeniy** [comment]
Hi,
 
Thank you for the update. We researched multiple ways of fixing the issue, but none of them were efficient. The logic which needs to be re-implemented is a part of our core mechanisms, which are common to the `Chart`, `Grid` and `Scheduler`. Any change there should be performed only after deep refactoring and accurate testing. 


This a is complex task which requires time and resources. We'll consider it in the future but cannot give you an estimate on it. In the meantime, we recommend using the solution I suggested before.
